torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py

```diff
@@ -334,6 +338,334 @@ def collect_bw_donated_buffer_idxs(
     return [fw_metadata.num_symints_saved_for_bw + i for i in fw_donated_buffer]
 
 
+@dataclasses.dataclass
+class InvokeSubgraphHopGraphs:
+    """
+    A data structure to hold all the information needed to partition the
+    `joint_hop_gm` and joint graph and the restitch the `new_fw_hop_gm` and
+    `new_bw_hop_gm` into the bigger `joint_gm`.
+    """
+
+    # To avoid re-partitioning subgraphs
+    partitioning_done: bool = False
+    old_num_fw_outputs: Optional[int] = None
+    old_num_fw_inputs: Optional[int] = None
+
+    new_fw_hop_gm: Optional[torch.fx.GraphModule] = None
+    new_bw_hop_gm: Optional[torch.fx.GraphModule] = None
+    # This includes (*fw_outs, *saved_tensors)
+    new_num_fw_outputs: Optional[int] = None
+
+
+def run_joint_graph_passes_on_hops(
+    joint_gm: torch.fx.GraphModule,
+    joint_inputs: Any,
+    aot_config: AOTConfig,
+) -> torch.fx.GraphModule:
+    """
+    This pass runs the joint graph passes on the HOP graph. In torch.compile, we
+    typically have many passes which work on the joint graph and then end with a
+    partitioner.
+
+
+    The partitioner part is quite mechanical to handle. HOP have their own
+    forward and backward graph. The process can be broken into following steps
+
+    1) Get a `joint_hop_gm` from the `fw_hop_gm` and `bw_hop_gm`
+    2) Run joint graph passes on the `joint_hop_gm` to get `new_fw_hop_gm` and `new_bw_hop_gm`
```

Comment:
basic mechanical q: don't the joint graph passes themselves just return an updated joint graph, that we later run the min cut partitioner on? Why do we need to manually stitch a forward and backward graph together at this point?

Reply:
oh, it might be because inductor gives us a `partition_fn()` that does both steps in one go (graph passes as well as invoking the min cut partitioner).

One question I have is for compile times: the min cut partition is known to be pretty slow, so invoking it redundantly on each individual subgraph and then again later on the entire (mixed) outer graph will be expensive.

Do you think it's worth splitting out those steps so we don't need to redundantly re-run the min-cut partitioner on each subgraph and stitch the graphs back together?

Reply:
The way current min-cut partitioner is - if we run on the main graph, it ignores the subgraphs. It is just a call_function node which has some meta info and does not really recurse. 

You could argue that min-cut partitioner should recurse, and perhaps there is some better implementation where recursion and stitching can happen together. But this is the simplest I could think of as of now (and also since this is just one HOP as of now, I did not want to boil the ocean by designing something very generic).